activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -220,9 +220,11 @@ def explain(arel, binds = [])
 
       # Executes the SQL statement in the context of this connection.
       def execute(sql, name = nil)
-        log(sql, name) do
-          ActiveSupport::Dependencies.interlock.permit_concurrent_loads do
-            @connection.query(sql)
+        with_connection_verify(::ActiveRecord::StatementInvalid) do
```

Comment:
If we are just trying to recover from lost connections, then we shouldn't need to retry all types of exceptions.